### PR TITLE
[AMDGPU] Implement variadic functions by IR lowering

### PR DESCRIPTION
**Pull request by JonChesterfield**
This is a mostly-target-independent variadic function optimisation and lowering pass. It is only enabled for AMDGPU in this initial commit.

The purpose is to make C style variadic functions a zero cost abstraction. They are lowered to equivalent IR which is then amenable to other optimisations. This is inherently slightly target specific but much less so than one might expect - the C varargs interface heavily constrains the ABI design divergence.

The pass is primarily tested from webassembly. This is because wasm has a straightforward variadic lowering strategy which coincides exactly with what this pass transforms code into and a struct passing convention with few cases to check. Adding further targets conventions is straightforward and elided from this patch primarily to simplify the review. Implemented in other branches are Linux X86, AMD64, AArch64 and NVPTX.

Testing for targets that have existing lowering for va_arg from clang is most efficiently done by checking that clang | opt completely elides the variadic syntax from test cases. The lowering produces a struct for each call site which can be inspected to check the various alignment and indirections are correct.

AMDGPU presently has no variadic support other than some ad hoc printf handling. Combined with the pass being inactive on all other targets landing this represents strict increase in capability with zero risk. Testing and refining will continue post commit.

In addition to the compiler tests included here, a self contained x64 clang/musl toolchain was constructed using the "lowering" instead of the systemv ABI and used to build various C programs like lua and libxml2.

### CODE REVIEW

File: clang/lib/CodeGen/Targets/AMDGPU.cpp
Comment (JonChesterfield) [inline clang/lib/CodeGen/Targets/AMDGPU.cpp:210]: This was subtle. Structs that aren't packed into integers and passed in registers fall through to default handling which sets CanBeFlattened, saying that it's OK to spread the struct across multiple arguments. This is then very difficult to reassemble robustly using the va_arg(x, type) interface - one needs to compute how type is likely to have been spread out across part of the call frame.

Noting that these values aren't being usefully passed in registers anyway, the `if (Variadic) {}` sets up call instructions that pass values by value (not byval) and declares that every value shall be exactly four byte aligned (including doubles, as that's something Matt suggested for amdgpu some time ago). This means the frame setup implementation and the case analysis for testing are very straightforward.

File: clang/test/CodeGenCXX/inline-then-fold-variadics.cpp
Comment (arsenm) [inline clang/test/CodeGenCXX/inline-then-fold-variadics.cpp:11]: Does this need REQUIRES: wasm-registered-target

File: clang/lib/CodeGen/Targets/AMDGPU.cpp
Comment (arsenm) [inline clang/lib/CodeGen/Targets/AMDGPU.cpp:108]: Can you split the clang AMDGPU ABI changes into a separate PR? The tests for this are also missing

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:238]:
```suggestion
    return FunctionType::get(FTy->getReturnType(), ArgTypes, /*IsVarArgs=*/false);
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (arsenm) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:782]: can do this in the constructor

File: clang/lib/CodeGen/Targets/AMDGPU.cpp
Comment (JonChesterfield) [reply clang/lib/CodeGen/Targets/AMDGPU.cpp:108]: Yes, I think I can do - the ABI change only affects variadic functions, which currently fatal_error anyway - but I think the C to IR tests will succeed as long as nothing calls va_arg and it stops before codegen.

**Comment by JonChesterfield** [comment]
I've spawned #93974 in case we can land the target independent part in parallel with me trying to debug the amdgpu/libc error. Maintaining different branches with different variations on what subset is implemented is confusing me quite a lot and leads to reviewers seeing essentially the same code repeatedly - I haven't found a way to represent base + N extensions in github without landing the base commit first.

File: libc/test/src/__support/arg_list_test.cpp
Comment (lntue) [inline libc/test/src/__support/arg_list_test.cpp:25]: Probably it's better to do:
```
template <typename T>
cpp::enable_if_t<cpp::is_integral_v<T>, void> dump(T s) {
  auto x = IntegerToString<T>(s);
  write_to_stderr(x.view());
  nl();
}
```
Then all the usages below could be changed to `dump(s.*);`

Comment (JonChesterfield) [reply libc/test/src/__support/arg_list_test.cpp:25]: Ah rubbish, the libc commit isn't supposed to be in this PR. That's a collection of local hacks trying to get libc to build and tell me useful things, which I've uploaded in error while patching Matt's comments. I'll work out how to remove that

Comment (JonChesterfield) [reply libc/test/src/__support/arg_list_test.cpp]: Now removed

File: clang/lib/CodeGen/Targets/AMDGPU.cpp
Comment (JonChesterfield) [reply clang/lib/CodeGen/Targets/AMDGPU.cpp:108]: Added the test to this PR and also split out #94083. Can land that subpatch first and rebase this for a reduction in complexity.

**Comment by JonChesterfield** [comment]
Finally managed to reproduce the libc failure. Thanks to Joseph for helping debug through the cmake. This patch as written was too optimistic about addrspacecast, a significantly more paranoid version behaves correctly (i.e. all the libc tests pass, this patch can be amended to enable the variadic ones on amdgpu). Libc tests don't seem to pick up changes to clang without a clean build (i.e. delete the whole build tree).

The memory error was on the dereference implied by va_arg. Valid addresses (read as void*) have a high bit pattern of 00002..... and invalid ones of fffff....., I'm unsure what that implies. I note that https://llvm.org/docs/AMDGPUUsage.html doesn't say much about the semantics of addrspace.

Trying to isolate that further. The error is definitely in the amdgpu-specific addrspace handling part though, the target agnostic part worked as intended.

**Comment by JonChesterfield** [comment]
I wish to ship this as-is. This patch is carefully constructed to be zero risk and landing it significantly improves the chances of zero-overhead varargs shipping. This revision passes the amdgpu libc tests. The tests are fragile to seemingly trivial changes to the IR generated which is obstructive to emitting nicer IR and some changes to the target abstraction I wish to make.

I have approximately isolated the libc / amdgpu memory error. I believe the amdgpu backend miscompiles constructs involving addrspacecast and passing structs to functions in a fashion which interacts badly with the lowering this pass chooses to do. That would explain why the transform has been easy and robust when implemented on x86, x64, nvptx, aarch64 and webassembly - and very prone to confusing memory errors and strange action at a distance on amdgpu. I note that this pass was originally made target-independent because I failed to get it running robustly on amdgpu alone.

Various previous revisions of this pass have passed the amdgpu libc tests and other similar revisions failed, I'd guess it's about 1/3 are "good" and 2/3 are "bad". I want to land this now, on a "good" revision, so we have something in tree that I can iterate on and feed into the ROCm dev branches, where I may have other colleagues or a working debugger available.

The pass core is straightforward and likely to be functionally correct. It isn't in the default pass pipeline so the only non-test time it'll run to begin with is for amdgpu. The lowering is exactly what webassembly does already and very close to nvptx. I'll implement x64 and aarch64 in parallel with chasing amdgpu oddities, both are mechanical exercises in case analysis and test coverage. I believe wasm needs no further changes.

Various things are suboptimal on amdgpu. Passing some things byref is better than the current everything inline layout. The extra parameter should be in (5), not (0). Codegen is noisier than it should be. Most permutations to the current scheme win "memory error" from HSA. I think these should be part of ongoing work after this lands and not a prereq.

**Comment by jhuber6** [comment]
I can confirm that it passes the tests against the `libc` targets, namely basic `stdarg.h` implementations and `sprintf`.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (efriedma-quic) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:387]: If you wanted to make this more general, you could scan the module for function declarations starting with llvm.va_start etc.  Not sure it's necessary.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (efriedma-quic) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:411]: There's a weird edge case here you might need to consider: a varargs call where the callee is a function that isn't varargs.  That's UB on its face... but some targets try to make it work.  And even on targets where it's purely UB, you still need to rewrite the call so the backend doesn't blow up.

Do you actually need the isIndirectCall() check?  Any call to a varargs function should have been rewritten at this point.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (efriedma-quic) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:624]: You can just delete this; it's never correct to mark the call "tail".

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (efriedma-quic) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:831]: What do you do if it's musttail?  I guess if it's a thunk, you can just forward the caller's varargs list.  Otherwise, the varargs argument list should be empty, I think?

File: llvm/utils/gn/secondary/llvm/lib/Transforms/IPO/BUILD.gn
Comment (efriedma-quic) [inline llvm/utils/gn/secondary/llvm/lib/Transforms/IPO/BUILD.gn:36]: There's a bot that updates the gn files automatically... so unless you're building this way, I'd suggest not touching this file.

**Review by efriedma-quic — COMMENTED** [review]
Left some review comments.  Please land https://github.com/llvm/llvm-project/pull/94083 first.

I agree it makes sense to land this in essentially its current form; it seems to be the right algorithm.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:387]: The addrspace qualifier on these intrinsics might be worth rolling back on the basis that their lowering is independent of addrspace on amdgpu and other targets don't use the addrspace forms. Even on amdgpu clang still emits the intrinsics using addrspace(0). I'll ping Alex about that sometime after this patch goes through.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:624]: Yes, the spurious comment is a symptom that there are far too many revisions of this in flight.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:831]: When optimising it's ignored. When lowering, a fatal error. Interesting point that must-tail call without passing any variadic arguments should be well formed. Currently it isn't, I've punted on what the empty va_list representation should be by putting a char in the frame. I'll think about how to fix that (it'll probably involve passing a target-specific representation of empty)

File: llvm/utils/gn/secondary/llvm/lib/Transforms/IPO/BUILD.gn
Comment (JonChesterfield) [reply llvm/utils/gn/secondary/llvm/lib/Transforms/IPO/BUILD.gn:36]: Nice, I did not know what. I am indeed not a BUILD.gn user.

**Comment by JonChesterfield** [comment]
Majority of review comments applied. Merged 94083 (with apologies to Matt) and rebased.

I claim the top level strategy of "turn ... into va_list then fix up known call sites" is a reasonable strategy for optimising variadic functions. The target specific quirks being abstracted over by the existing va_list interface is pleasing.

The target abstraction here is not totally what I want it to be and there are some simplifying assumptions (e.g. the C calling conv check probably needs to be under target-specific, variadic invokes exist in libc++) that won't survive for long.

Still, I think this is a worthwhile step forward. It's the machinery needed to render variadic functions zero cost at compile time, and gives back ends with freedom over their ABI the option to totally ignore variadic functions. @bigboze this is the pass I looked for during the IPU bring up - it would have been great to say "a void *, four byte alignment seems fine" and cross "implement varargs" off the backlog in half an hour, instead of hand implementing the thing in sdag. Shipping this would also mean successfully avoiding implementing varargs in sdag and globalisel for amdgpu, and whatever nvptx codegen uses, so there's a definite code reuse theme from the IR strategy.

My near immediate future involves trying to work out what the addrspace/struct/memory-error weirdness on amdgpu is but I'm also looking forward to the variadic wrapper functions in my headers on x64 disappearing at O1 time.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (Pierre-vh) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:625]:
```suggestion
  if (ABI->vaListPassedInSSARegister())
    Args.push_back(Builder.CreateLoad(ParameterType, VaListInstance));
  else
    Args.push_back(Builder.CreateAddrSpaceCast(VaListInstance, ParameterType));
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (Pierre-vh) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:466]:
```suggestion
  for (User *U : make_early_inc_range(VariadicWrapper->users())) {
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (Pierre-vh) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:744]:
```suggestion
    else  if (IsByVal) 
      Frame.memcpy(Ctx, FrameFieldType, SourceValue, UnderlyingSize);
    else 
       Frame.store(Ctx, FrameFieldType, SourceValue);
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (Pierre-vh) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:932]: Not critical at all right now (maybe put a TODO?), but at some point it might be better to put the VariadicABIInfo class in the header, and just require it as a parameter. Then all the target-specific stuff can move into `Target` and be provided by the `TargetMachine`

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (Pierre-vh) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:944]: Leftover todo?

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (Pierre-vh) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:1005]:
```suggestion
VariadicABIInfo::create(const Triple &T) {
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (Pierre-vh) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:381]:
```suggestion
  for (Function &F : make_early_inc_range(M))
```

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (Pierre-vh) [inline llvm/lib/Transforms/IPO/ExpandVariadics.cpp:364]:
```suggestion
  Triple TT(M.getTargetTriple());
```
(TT is a bad name, but I prefer that than having a variable named like a type)

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:944]: Sort of todo, sort of comment to reviewers. Removed it.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:99]: I'm pretty sure I've been told to move const to the other side by other reviewers. Someone should check which side is the one that works in more cases and teach clang format about it.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (JonChesterfield) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:364]: I've gone with anonymous. I called it triple, but that Violates Naming Conventions, so prefixed llvm:: to remove the ambiguity. I wonder how many variables we end up with called V as a result of snake_case_rejection.

**Comment by JonChesterfield** [comment]
Thanks @Pierre-vh, I think those are all applied. Also removed llvm:: from some other places, and remembered that llvm doesn't like the if () {braces} which are so useful for git merge not introducing bugs so removed the majority of those as well. Feel confident enough after reading through to mark it green? All it can do is take variadic functions from hard errors to probably working.

**Comment by JonChesterfield** [comment]
An offline suggestion from Pierre is that this should early-exit if there are no variadic functions in the module. That's a good thing, I'd like to consider it another of the increase-complexity-for-decreased-compile-time to implement after something has landed.

**Comment by jhuber6** [comment]
> An offline suggestion from Pierre is that this should early-exit if there are no variadic functions in the module. That's a good thing, I'd like to consider it another of the increase-complexity-for-decreased-compile-time to implement after something has landed.

I thought any use of varargs would introduce LLVM intrinsics. Shouldn't it be trivial to look if any of those exist in the module?

**Comment by JonChesterfield** [comment]
We can check for the va_start which is a good approximation, but there can be a variadic function pointer from another module and the only place that would show up is a call instruction, and the only way to find that is enumeration. Amdgpu can somewhat cheat using the whole.program assumption but it's sketchy.

File: libc/test/CMakeLists.txt
Comment (jhuber6) [inline libc/test/CMakeLists.txt:11]: Can we precommit or move this to a separate patch

**Review by jhuber6 — COMMENTED** [review]
Overall I think this is good. Address the existing comments and I think we should be able to land it. Potentially we should be able to check for the existence of `va_start` in the module to early-exit like you said, which will keep functional changes to a minimum.

FWIW if you use `clangd` with `clang-tidy` it should let you automatically fix the LLVM style issues.

**Comment by JonChesterfield** [comment]
Early exit on lack of va_start will be incorrect in the lowering case, which is the only one enabled by default. I believe existing comments are all addressed. Precommit the cmake diagnostic tweak sounds good, would you like to land that?

File: libc/test/CMakeLists.txt
Comment (JonChesterfield) [reply libc/test/CMakeLists.txt:11]: This is helpful for me working out why libc isn't running tests but otherwise unrelated to this PR, I'll drop it from the change and you can apply or discard it separately as you prefer.

**Comment by jhuber6** [comment]
> Early exit on lack of va_start will be incorrect in the lowering case, which is the only one enabled by default. I believe existing comments are all addressed.

Figured if there's no `va_start` there's nothing for the pass to do anyway.

> Precommit the cmake diagnostic tweak sounds good, would you like to land that?

Sure

**Comment by JonChesterfield** [comment]
If there's no va_start and we're optimising there's no work to do. Otherwise when lowering we still transform declarations (which have no basic blocks) and calls to unknown pointers, so that separate compilation works. That's why knowing it's the whole program would be useful.

File: libc/test/CMakeLists.txt
Comment (jhuber6) [reply libc/test/CMakeLists.txt:11]: Done

File: llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp
Comment (Pierre-vh) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:996]: Please add a kill switch `cl::opt` for the pass in the AMDGPUTargetMachine before landing, in case it has unforeseen issues. It'd be easier to default-disable while fixing, rather than revert a big patch :)

**Review by Pierre-vh — APPROVED** [review]
LGTM if a kill switch is added, I would wait a day more before landing in case anyone else has concerns

File: llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp
Comment (JonChesterfield) [reply llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:996]: I'm also paranoid enough to want that, it's '--expand-variadics-override=disable' from the command line or changing the ::Lowering in that line to ::Disable. Commenting out the addPass also kills it, but would cause the lit pipeline tests to fail. I'll put a comment above that pass saying this.

**Comment by JonChesterfield** [comment]
I've been considering the relative merits of waiting a day for further comments and landing on a Friday before I'm out of the office for a week. I think the better play is land now, pick up the merge into internal CI testing, and have tomorrow to pick up any further comments and deal with unanticipated problems.

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (arsenm) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp]: const always to the left. East const is weird

File: llvm/lib/Transforms/IPO/ExpandVariadics.cpp
Comment (arsenm) [reply llvm/lib/Transforms/IPO/ExpandVariadics.cpp:386]: That's not the nature of the address space, and clang should be emitting these in the natural stack address space

**Comment by JonChesterfield** [comment]
It looks like the libc cmake change was not precise enough, nvtpx is trying to run libc.src.stdio.printf_core.printf_main.dir and shouldn't be. Looking at that now

**Pull request merged**
